swarm/src/behaviour/peer_addresses.rs

```diff
@@ -0,0 +1,306 @@
+use crate::behaviour::FromSwarm;
+use crate::{DialError, DialFailure, NewExternalAddrOfPeer};
+
+use libp2p_core::Multiaddr;
+use libp2p_identity::PeerId;
+
+use lru::LruCache;
+
+use std::collections::HashSet;
+use std::num::NonZeroUsize;
+
+/// Struct for tracking peers' external addresses of the [`Swarm`](crate::Swarm).
+#[derive(Debug)]
+pub struct PeerAddresses(LruCache<PeerId, HashSet<Multiaddr>>);
+
+impl PeerAddresses {
+    /// Creates cache with capacity of `cache_size`.
+    pub fn new(cache_size: NonZeroUsize) -> Self {
+        Self(LruCache::new(cache_size))
+    }
+
+    /// Feed a [`FromSwarm`] event to this struct.
+    ///
+    /// Returns whether the event changed peer's known external addresses.
+    pub fn on_swarm_event(&mut self, event: &FromSwarm) -> bool {
+        match event {
+            FromSwarm::NewExternalAddrOfPeer(NewExternalAddrOfPeer { peer_id, addr }) => {
+                if let Some(peer_addrs) = self.0.get_mut(peer_id) {
+                    let addr = prepare_addr(peer_id, addr);
+                    peer_addrs.insert(addr)
+                } else {
+                    let addr = prepare_addr(peer_id, addr);
+                    self.put(*peer_id, std::iter::once(addr));
+                    true
+                }
+            }
+            FromSwarm::DialFailure(DialFailure {
+                peer_id: Some(peer_id),
+                error: DialError::NoAddresses,
+                ..
+            }) => self.0.pop(peer_id).is_some(),
```

Comment:
Sorry that I am just noticing this now but I don't think this makes any sense! `DialError::NoAddresses` means we didn't have any addresses to complete the dial. Purging the cache in this situation doesn't make any sense as we seem to not have any addresses anyway :)

Instead of `NoAddresses`, we should use the `Transport` variant and call `self.remove` for each address in there.

Reply:
:heavy_check_mark: